### ISSUE TITLE
Handle Javac Optimisation of String Concatenations

### ISSUE THREAD

**Issue by ykoitzsch** [neow3j/neow3j#781]
**Description of the bug**
It appears that when I am trying to concatenate two Strings, when one is from the storage, then the compiler complains with the following message:

```
Inherited methods that are not specifically implemented are not supported. Implement the method 'toString' without a 'super' call to the class Object to use it.
io.neow3j.compiler.CompilerException: Inherited methods that are not specifically implemented are not supported. Implement the method 'toString' without a 'super' call to the class Object to use it.
        at io.neow3j.compiler.Compiler.throwIfObjectIsOwner(Compiler.java:603)
        at io.neow3j.compiler.Compiler.handleInsn(Compiler.java:584)
        at io.neow3j.compiler.NeoMethod.convert(NeoMethod.java:481)
        at io.neow3j.compiler.converters.MethodsConverter.handleUncachedMethodCall(MethodsConverter.java:224)
```

**To Reproduce**
1. create new smart contract
1.  try to concatenate two strings: map.getString(key) + "foo";
2. or: map.get(key).toString() + "foo"
3. compile 

**Expected behavior**
It should be possible to concatenate two strings like this without compiling issues.

**Platform and Software**
 - neow3j 3.17

**Comment by gsmachado** [comment]
@mialbu is this issue somehow related to https://github.com/neow3j/neow3j/issues/760?

**Comment by gsmachado** [comment]
also, @mialbu, if we don't have a test for this particular case pointed out by @ykoitzsch, we should add it to our int tests suite.

**Comment by mialbu** [comment]
I cannot reproduce this with 3.17.0 nor with the most recent not yet released neow3j code.

@ykoitzsch, could you provide more context, such as the `map` initialization and where you're getting the `key` from (e.g., from method param, static value of the contract, ...)?

**Comment by ykoitzsch** [comment]
Here is an example:
```
public class Dummy {

    private static final StorageContext ctx = Storage.getStorageContext();

    private static final StorageMap sugarValues = new StorageMap(ctx, (byte) 104);

    public static String foo() {
        return sugarValues.getString("key") + "foo";
    }

}
```

**Comment by mialbu** [comment]
> Here is an example:
> 
> ```
> public class Dummy {
> 
>     private static final StorageContext ctx = Storage.getStorageContext();
> 
>     private static final StorageMap sugarValues = new StorageMap(ctx, (byte) 104);
> 
>     public static String foo() {
>         return sugarValues.getString("key") + "foo";
>     }
> 
> }
> ```

I can still not reproduce this on my machine. When there's a value stored at the key `key`, the VM halts and returns the expected stack. If no value is set, the VM faults with `Specified cast is not valid.`, which is expected to happen.

Reading the compiler exception, could it be that somewhere in your code you use the `toString` method on a custom object (e.g., a @Struct annotated class) that gets compiled as well? Or does the compilation fail when you compile only exactly this dummy class? I'm asking this, because we throw this `CompilerException` when the `toString` method of the class `Object` is called, which should nowhere be the case in this dummy class.

**Comment by ykoitzsch** [comment]
It fails only when I compile this class.  Removing the concatenation resolves the issue, and I can compile it successfully.
Here is my compiling code:

```
public static CompilationUnit run(Class<?> contract)
                        throws Throwable {
                CompilationUnit res = new Compiler().compile(
                                contract.getCanonicalName());
                Path filePath = Paths.get("build", "neow3j");
                filePath.toFile().mkdirs();
                ContractUtils.writeNefFile(res.getNefFile(),
                                res.getManifest().getName(), filePath);
                ContractUtils.writeContractManifestFile(res.getManifest(),
                                filePath);
                return res;
        }
```

**Comment by ykoitzsch** [comment]
But seeing that it works for you, the issue might be somewhere on my side. I'll test it in an empty project and report back.

**Comment by ykoitzsch** [comment]
cloned this repo: https://github.com/neow3j/neow3j-boilerplate-contracts and tested it. Same error message.

**Comment by ykoitzsch** [comment]
Okay, I might have figured out the issue. I am used to running things directly in VSC by clicking the run button. Running it via the command line works just fine.

**Comment by ykoitzsch** [comment]
So, I guess the issue can be closed. Sorry for the troubles.

**Comment by gsmachado** [comment]
> Okay, I might have figured out the issue. I am used to running things directly in VSC by clicking the run button. Running it via the command line works just fine.

Interesting.

I would be interested to know your EXACT environment:
- VSCode version
- Java version (globally)
- VSCode Java plugins name & plugins versions

The problem might be as follows: VSCode Java plugin might ignore configurations in the Gradle build file, and, maybe it compiles the `.class` based on a version that is not Java 1.8. It probably uses the version from your machine (which might be Java 11 or 13, or something like this).

**Comment by gsmachado** [comment]
> Okay, I might have figured out the issue. I am used to running things directly in VSC by clicking the run button. Running it via the command line works just fine.

@ykoitzsch can you please tell me which command line did you use? Was it `./gradlew clean neow3jCompile`?

**Comment by ykoitzsch** [comment]
I have multiple contracts in my project, so I use the application plugin and added the following lines to the gradle.build:

```
ext {
   javaMainClass = "io.neocandy.deployment.candyclash.CUCompile"
}

application {
    mainClassName = javaMainClass
}
```
then I run the class with ./gradlew run. 
So since I can't run main classes with VSC successfully I now just replace the javaMainClass with whatever I want to run. It's not perfect but at least works for now :D

**Comment by gsmachado** [comment]
> I have multiple contracts in my project, so I use the application plugin and added the following lines to the gradle.build:
> 
> ```
> ext {
>    javaMainClass = "io.neocandy.deployment.candyclash.CUCompile"
> }
> 
> application {
>     mainClassName = javaMainClass
> }
> ```
> 
> then I run the class with ./gradlew run. So since I can't run main classes with VSC successfully I now just replace the javaMainClass with whatever I want to run. It's not perfect but at least works for now :D

Alright. Gotcha.

Now, the last question: do you have `sourceCompatibility` and `targetCompatibility` on your `build.gradle`? If yes, please paste it here.

**Comment by ykoitzsch** [comment]
sourceCompatibility = 1.8
targetCompatibility = 1.8

**Comment by gsmachado** [comment]
@ykoitzsch alright. We probably know the exact issue.

One more question: can you tell me, if you open a terminal and type `java -version`, what do you get?

Also, can you check which Java HOME (or runtime) is set in your VSCode? Probably you can find that on your `settings.json` file.

**Comment by gsmachado** [comment]
> sourceCompatibility = 1.8 targetCompatibility = 1.8

Just for our info: using VSCode (RedHat java plugin), this would be nice and avoid these kind of problems: https://github.com/redhat-developer/vscode-java/issues/982

**Comment by ykoitzsch** [comment]
openjdk version "1.8.0_332"
OpenJDK Runtime Environment Corretto-8.332.08.1 (build 1.8.0_332-b08)
OpenJDK 64-Bit Server VM Corretto-8.332.08.1 (build 25.332-b08, mixed mode)

and it looks like I have no java.home set in my settings.json

**Comment by mialbu** [comment]
@ykoitzsch I could reproduce now. Javac is run with different compile options in VSCode than in Intellij. This includes an optimisation concerning the use of the `StringBuilder` class. It ultimately breaks the compilation to Neo byte code when an an optimised concatenation was produced by the Java compiler.

I'll add this issue to our backlog for now and we'll address it in the future. As for now, until we've fixed this, please use the gradle plugin to compile when developing with VSCode.

If you're interested in more details, check out [this](https://medium.com/javarevisited/java-compiler-optimization-for-string-concatenation-7f5237e5e6ed) and [this](https://stackoverflow.com/questions/24614762/does-javacs-stringbuilder-optimisation-do-more-harm-than-good?noredirect=1&lq=1).

### Additional Info
The following contract class fails to compile using VSCode:

```java
class StringConcatenation {
        private static final StorageContext ctx = Storage.getStorageContext();
        public static String getSomething(String key) {
            return Storage.getString(ctx, key) + "foo";
        }
}
```

In the following I've listed the pattern of the compiled asm class when using either Intellij or VSCode.

**Parts of the asmClass instructions based on compilation using Intellij IDE**

```
TypeInsnNode StringBuilder 187
InsnNode 89
MethodInsnNode StringBuilder <init> 183
FieldInsnNode StorageContext ctx 178
VarInsnNode 25
MethodInsnNode Storage getString 184
MethodInsnNode StringBuilder append 182
LdcInsnNode 18
MethodInsnNode StringBuilder append 182
MethodInsnNode StringBuilder toString 182
InsnNode 176
```

**Parts of the asmClass instructions based on compilation using VSCode IDE with RedHat Java extension**
```
TypeInsnNode StringBuilder 187
InsnNode 89
FielInsnNode StorageContext ctx 178
VarInsnNode 25
MethodInsnNode Storage getString 184
MethodInsnNode String valueOf 184
MethodInsnNode String <init>
LdcInsnNode 18
MethodInsnNode StringBuilder append 182
MethodInsnNode StringBuilder toString 182
InsnNode 176
```

**Comment by csmuller** [comment]
We will adapt the compiler to support this separate behaviour of the Java compiler.

**Comment by mialbu** [comment]
Hey @ykoitzsch, we've fixed this and I've published a snapshot for it: `3.18.1-SNAPSHOT` if you want to try it right away. We're planning on another release by end of July or in the beginning of August when it'll be included. Refer to https://github.com/neow3j/neow3j/blob/main/DEVELOPMENT.md for how to use a snapshot version of neow3j.

**Comment by gsmachado** [comment]
> Hey @ykoitzsch, we've fixed this and I've published a snapshot for it: `3.18.1-SNAPSHOT` if you want to try it right away. We're planning on another release by end of July or in the beginning of August when it'll be included. Refer to https://github.com/neow3j/neow3j/blob/main/DEVELOPMENT.md for how to use a snapshot version of neow3j.

@ykoitzsch it would be great if you could try this before we release it. 🙏 

Let us know the outcome. 😄

**Comment by ykoitzsch** [comment]
I tested it, and it compiles just fine now @mialbu  @gsmachado :)
